sycl/doc/extensions/proposed/sycl_ext_oneapi_append_and_shift.asciidoc

```diff
@@ -0,0 +1,128 @@
+The proposed extension of sycl::shift_group_left takes two values, the "current" value `c` and the "next" value `n`. If `id+delta` is less than the group's linear size, the function returns `c` of the work-item with group local id `id+delta`. If `id+delta` is greater or equal than the group's linear size, the function returns `n` of the work-item with the group local id `id+delta-sg_size`. Thus, in all cases the return value is defined and valid.
+Similarly, the proposed extension of `sycl::shift_group_right` returns `c` of the work-item with group local id `id-delta`, if `id-delta` is greater or equal 0 and `n` of the work-item with group local id `sg_size+(id-delta)`, otherwise.
+While this feature can easily be reproduced manually, is indicated below in the section "possible implementation", an extension is feasible since the underlying `pass:[__]spirv_SubgroupShuffleDownINTEL` and `pass:[__]spirv_SubgroupShuffleUpINTEL` already take the proposed two values and support the required capabilities.
+Note that there are multiple alternatives for the API suggested below ("append_and_shift_group_*"). For example, 
+
+
+``` 
+template <typename Group, typename T> 
+shift_group_left(Group g, T c, Group::linear_id_type delta=1, T n = 0);
+```
+or
+
+```
+template <typename Group, typename T> 
+shift_group_left_app(Group g, T c, T n, Group::linear_id_type delta=1);
+```
+
+
+== Specification
+
+=== Feature test macro
+This extension provides a feature-test macro as described in the core SYCL specification. An implementation supporting this extension must predefine the macro `SYCL_EXT_ONEAPI_APPEND_AND_SHIFT` to one of the values defined in the table below. Applications can test for the existence of this macro to determine if the implementation supports this feature, or applications can test the macro's value to determine which of the extension's features the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+	
+|1
+|The APIs of this experimental extension are not versioned, so the feature-test macro always has this value.
+|===
+
+=== API
+
+```
+template <typename Group, typename T>
+T sycl::ext::oneapi::append_and_shift_group_left(Group g, T c, T n, Group::linear_id_type delta=1);
+```
+
+_Constraints_: Available only if `std::is_same_v<std::decay_t<Group>, sub_group>` is true and `T` is a trivially copyable type.
+
+_Preconditions_: `delta` must be the same value for all work-items in the group.
+
+_Returns_: The value `c` of the work-item with group local id `id+delta`. If `id+delta` is greater or equal than the group's linear size, the function returns `n` of the work-item with the group local id `id+delta-sg_size`.

```

Comment:
`delta` may be larger than the group's linear size but then it would behave as the current implementation of shift_group_* applied to the `next` value with a shift value of (`delta`-`group's linear size`). I.e. no values from `current` would be taken. It may make sense to give a warning in that case? I think it would be sufficient as a disclaimer in the returns section stating which return values may be undefined in that case.

Regarding your comment above: I am not quite sure if I understand you correctly. But I think you mean that we could do a cyclic implementation on a single input and the user then has to overwrite the `current` value in the first `delta` number of work items with the `next` value before calling the function with just the `current` value, right? That would work as well. The reason I took current and next as input was due to the underlying spirv function, which takes the two inputs.

Reply:
> It may make sense to give a warning in that case?  I think it would be sufficient as a disclaimer in the returns section stating which return values may be undefined in that case.

The spec should probably say something like:

> The `delta` may be greater than or equal to the group’s linear size, but the value returned in this case is unspecified.